### ISSUE TITLE
IOperation API support for BoundDynamicMemberAccessExpression

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#20117]
API Proposal:

```
    public interface IDynamicMemberReferenceExpression : IOperation
    {
        /// <summary>
        /// Instance receiver.
        /// </summary>
        IOperation Instance { get; }

        /// <summary>
        /// Referenced member.
        /// </summary>
        string MemberName { get; }

        /// <summary>
        /// Type arguments.
        /// </summary>
        ImmutableArray<ITypeSymbol> TypeArguments { get; }
    }
```

**Comment by mavasani** [comment]
@jinujoseph We can discuss this today.

**Comment by jinujoseph** [comment]
**Design Team Decision**

Looks great, lets go with this.

**Comment by 333fred** [comment]
We should also replace `ILateBoundMemberReferenceExpression` with this. However, we're going to need a small addition to the interface for this. Consider the following VB code:

```
Option Strict Off

Module Module1
    Sub Main()
        Dim x As Object = New List(Of Integer)
        Try
            fun(x)
        Catch e As Exception
            Console.WriteLine(e)
        End Try
    End Sub

    Sub fun(Of X)(ByVal a As List(Of X))
    End Sub

End Module
```

In this code, `fun` is late-bound because the type arguments can't be inferred, but there is no receiver type. We need to add a field to our class to represent the containing type, which will always be null C#:

```C#
        /// <summary>
        /// The containing type of this expression. In C#, this will always be null.
        /// </summary>
        ITypeSymbol ContainingType { get; }
```

Additionally, we will need to amend the `Instance` comment:
```
        /// <summary>
        /// Instance receiver. In VB, this can be null.
        /// </summary>
        IOperation Instance { get; }
```

Fyi @dotnet/analyzer-ioperation @jinujoseph we should bring this up in design discussion this Thursday.

**Comment by jinujoseph** [comment]
**Design Team decision**

The proposal seems reasonable. We don't need separate types for LateBound.. and Dynamic and we'd have the containing Type needed for VB but C# wouldn't set them. We should try restrict to may be containing module and it's only for VB to use if the instance is null.

**Comment by 333fred** [comment]
I looked into narrowing the scope. Unfortunately, it doesn't seem possible, because of this scenario:
```VB
Option Strict Off
Imports System.Collections.Generic

Module Module1
    Sub Main()
        Dim x As Object = New List(Of Integer)()
        Dim c1 As New C1
        c1.fun(x)'BIND:"c1.fun(x)"
    End Sub
 
    Class C1
        Sub fun(Of X)(ByVal a As List(Of X))
        End Sub
    End Class
```

Here, the containing type is `Module1.C1`.

**Comment by CyrusNajmabadi** [comment]
@333fred Sorry, not sure what you're referring to.

**Comment by 333fred** [comment]
Cyrus and I talked offline. What we're going to is check to see if the Containing Type is different from the type of the receiver. If these types are not the same, then ContainingType will be set. Otherwise, ContainingType will be null.

**Comment by 333fred** [comment]
It will also be set if the Receiver is Nothing.

**Comment by jinujoseph** [comment]
fixed via #20966
